new rds_snapshot module - split of RDS module

michael-dev2rights opened the pull request:
##### SUMMARY

This PR contains a new rds_snapshot module. It is in fact a reworking of https://github.com/ansible/ansible/pull/25646 and https://github.com/ansible/ansible/pull/19772

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

rds_snapshot module

##### ANSIBLE VERSION

```
ansible 2.4.0 (mdd_rds_split_snapshot c2f27882c8) last updated 2017/07/10 15:26:05 (GMT +100)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/mikedd/dev/ansible/ansible-add-modules/library']
  ansible python module location = /home/mikedd/dev/ansible/ansible-add-modules/lib/ansible
  executable location = /home/mikedd/dev/ansible/ansible-add-modules/bin/ansible
  python version = 3.5.2 (default, Nov 17 2016, 17:05:23) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION

This is a _two time_ updated version of https://github.com/ansible/ansible/pull/19772 from Willthames  now split out so that  this PR only contains the rds_snapshot module

> Split the existing rds module into three separate modules, and make the modules idempotent (remove the command argument).
>  The new modules are expected to be almost entirely backward compatible with the old module, other than the command change - it should be easy to update existing code to use the new modules.
> The new modules are entirely boto3 - support for boto has been dropped for the new modules.

Willthames stated that he's not going to have much time to work on this (though hopefully he'll be able provide a little support and e.g. review the code) so I'm now proposing to put this forward myself as multiple separate PRs

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

Commit:
Provide all returned detail as facts

Commit:
Add rds_snapshot module

mikedlr committed:
rds_instance module - fix various parameter handling cases from trying to run integration test cases

mikedlr committed:
rds_instance - fixes & tests to ensure wait instance and changed state work

Commit:
Tidy up flake8 violations in module_utils.rds

mikedlr committed:
rds modules and tests - pep8 / flake8 error fixes

mikedlr committed:
rds modules and tests - further pep8 and sanity fixes including docs

mikedlr committed:
rds instance - validation fixes with nasty hack for from_future

mikedlr committed:
rds instance - continued cleanup to appease the CI gods

mikedlr committed:
module utils - move rds.py into recently agreed aws subdirectory

mikedlr committed:
rds_snapshot module - move rds import into recently agreed aws subdirectory

mikedlr committed:
rds utils tests - first version as a design document

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+  type: dict
+'''
+
+import botocore
```
Wrap in try/except please.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+
+# import module snippets
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import ec2_argument_spec, get_aws_connection_info, boto3_conn
```
import HAS_BOTO3 too and add a check for it in main()

s-hertel commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,125 @@
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'DBInstanceNotFound':
+            return None
+        else:
```
Don't need to specify the else. It will raise of its own accord.

s-hertel commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,125 @@
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'DBSnapshotNotFound':
+            return None
+        else:
```
Same as above.

s-hertel commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,125 @@
+                    before[k] = self.data.get(k)
+                    after[k] = params.get(k)
+        try:
+            old_port = self.data.get("endpoint")["port"]
```
Instead of wrapping in try/excepts you could do:
`old_port = self.data.get("endpoint", {}).get("port")`
Same with new_port.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+        # Temporary until all the rds2 commands have their responses parsed
+        if resource.name is None:
+            module.fail_json(
+                msg="There was a problem waiting for RDS snapshot %s" %
```
Can get rid of unnecessary line breaks.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+        if e.response['Error']['Code'] == 'DBSnapshotNotFound':
+            module.exit_json(changed=True)
+        else:
+            module.fail_json(msg=e.message)
```
Can you add traceback to the exception handling please?

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+            snapshot = RDSSnapshot(resource['DBSnapshot'])
+            changed = True
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg=e.message)
```
See https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-2
Can update most of the exception handling. Helpful for future debugging! :)

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+        module.fail_json(msg="Region not specified. Unable to determine region from EC2_REGION.")
+
+    # connect to the rds endpoint
+    conn = boto3_conn(module, 'client', 'rds', region, **aws_connect_params)
```
Could raise botocore.exceptions.NoCredentialsError or botocore.exceptions.ProfileNotFound

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+            instance_name=dict(),
+            wait=dict(type='bool', default=False),
+            wait_timeout=dict(type='int', default=300),
+            tags=dict(type='dict', required=False),
```
Can get rid of required=False.

mikedlr committed:
rds utils - rewriting for new functions to handle rds data better

mikedlr committed:
rds snapshot module - fixes responding to shertel review including move to new AnsibleAWSModule

mikedlr committed:
rds utils - clean out remaining old object definitions

mikedlr committed:
rds utils - get through to having working unit tests

Commit:
Simplify rds common module and add tests

Rely on boto3 structure for RDS snapshot and remove RDSSnapshot class.
Add tests for RDS snapshot
Ensure IAM configuration is correct for tests

willthames commented:
@mikedlr I pushed a new commit on to your branch that fixes the automated tests and @s-hertel's issues, and adds integration tests, IAM policies and refactors the rds common module somewhat. 

This should be a good starting point for fixing the other three modules too, hopefully I can address some of the other concerns tomorrow.

The tests might not quite pass - I got it mostly working before bedtime, apologies if I missed anything (mostly changing RDSSnapshot attributes to boto3 rds dictionary elements)

mattclay commented:
In addition to the failures noted by the bot, the integration tests are also failing with:

```
{
"failed": true, 
"msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined.The error was: 'aws_region' is undefined\n\nThe error appears to have been in '/root/ansible/test/integration/targets/rds_snapshot/tasks/main.yml': line 30, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  always:\n    - name: destroy DB\n      ^ here\n\nexception type: <class 'ansible.errors.AnsibleUndefinedVariable'>\nexception: 'aws_region' is undefined"
}
```

mikedlr committed:
rds utils - aggressive fixing of pep8 and other issues - unit tests work on python3 as well

mikedlr committed:
Revert "Simplify rds common module and add tests"

This reverts commit b268b34ced66e941abf1b47f402a60c3ef0b3d1d.

For now - to be discussed with Will.
- Integration tests already exist in separate PR and do more
- Simplification is also being done in separate
- Plan to have separate functions for converting instance to facts

mikedlr committed:
from willthames - Ensure IAM configuration is correct for tests

mikedlr committed:
rds utils - add various constant data lists - rename functions and various minor fixes

mikedlr committed:
rds_snapshot module - begin refactoring to work with new rds utils

mikedlr committed:
rds_snapshot module - more updates for new rds utils - fix delete

mikedlr commented:
@willthames could you close this please;  it's replaced by the [newer PR](https://github.com/ansible/ansible/pull/30746)
